xarray/core/dataarray.py

```diff
@@ -260,7 +260,8 @@ def __init__(
             # try to fill in arguments from data if they weren't supplied
             if coords is None:
 
-                if isinstance(data, DataWithCoords):
+                if isinstance(data, DataArray):
+                    coords = data.coords
                     coords = getattr(data, 'coords', None)
```

Comment:
Sorry, this line should be removed -- suggestions in GitHub comments can't replace multiple lines.

Reply:
Yeah, Github's diffs give often me a headache. This is what we want, correct?
```python
            if coords is None:

                if isinstance(data, DataArray):
                    coords = data.coords
                if isinstance(data, pd.Series):
                    coords = [data.index]
                elif isinstance(data, pd.DataFrame):
                    coords = [data.index, data.columns]
                elif isinstance(data, (pd.Index, IndexVariable)):
                    coords = [data]
                elif isinstance(data, pdcompat.Panel):
                    coords = [data.items, data.major_axis, data.minor_axis]
```

Reply:
yes... probably could convert `if isinstance(data, pd.Series):` to use an `elif`